torch/fx/split.py

```diff
@@ -0,0 +1,188 @@
+
+    submod_graph = Graph()
+    submod_delegate = DefaultDelegate(target_module, submod_graph)
+
+    # Map Node in the original Graph to the node in the new graph
+    remap_table = {}
+    for block_input in block_inputs:
+        remap_table[block_input] = submod_delegate.placeholder(block_input.name)
+
+    for node in block_nodes:
+        def arg_transform(n : Node):
+            return remap_table[n]
+
+        # We need to fixup all qualfied names to strip off the base
+        # qualname, since now all qualnames in the new Module should
+        # be wrt that new module
+        def qualname_transform(qualname : str) -> str:
+            return qualname.replace(f'{target_qualname}.', '')
+
+        submod_graph_node = submod_graph.node_copy(node, arg_transform, qualname_transform)
+
+        # Keep track of created output values
+        remap_table[node] = submod_graph_node
+
+    # Register output
+    submod_graph.output(remap_table[block_outputs[0]])
+    submod_module = GraphModule(target_module, submod_graph)
+
+    # ===== Stage 3: Create new base GraphModule with submodule nodes replaced
+    #       with a callsite to that new submodule =====
+
+    base_graph : Graph = Graph()
+    base_delegate : DefaultDelegate = DefaultDelegate(mod, base_graph)
+    base_remap_table : Dict[Node, Node] = {}
+
+    callsite_inserted = False
+    for orig_node in mod.graph.nodes:
+        def arg_transform(n : Node):
+            return base_remap_table[n]
+        if orig_node in block_defs:
+            if not callsite_inserted:
+                args = tuple(arg_transform(n) for n in block_inputs)
+                base_remap_table[block_outputs[0]] = base_delegate.create_node(
+                    'call_module', target_qualname, args, {}, block_outputs[0].name,
+                    orig_node.module_qualname)
+                callsite_inserted = True
+            # Fallthrough. For each node that's a def in the subgraph,
+            # just elide copying it.
+        else:
+            base_remap_table[orig_node] = base_graph.node_copy(orig_node, arg_transform)
+
+    # Register output
+    assert isinstance(mod.graph.result, Node)
+    base_graph.output(base_remap_table[mod.graph.result])
+
+    # A bit of explanation is in order here. Since we're inserting a callsite
+    # to a submodule that may not exist on the original root (as is the case
+    # with stateless submodules) we may need to construct "fake" submodules
+    # up to and including the new submodule we've just constructed. This
+    # function recursively traverses through the target_qualname_atoms to make
+    # sure we have instantiated Modules up to and including the new submodule.
+    # In the base case, we construct the final GraphModule with the new graph,
+    # containing the call to the new `target` module. On the return path, we
+    # return the hierarchy within the passed-in `mod` back to its original state
+    def construct_graphmodule_with_correct_submod(
```

Comment:
Is this the same thing? I am having trouble following the recursion:
```
cur = root
*prefix, last = target_qual_name.split('.')
for atom in prefix:
     if not hasattr(cur, atom):
       setattr(cur, atom, torch.nn.Module()
     cur = getattr(cur, atom)
setattr(cur, last, GraphModule(mod, base_graph))
```

Reply:
No, this version undoes the `setattr(root_submod, curr_atom, torch.nn.Module())` for each atom in the qualname. The recursion is using the callstack as a data structure to keep track of those

Reply:
I don't understand this then: What is the effect of this function if it just undoes the construction of submodules?

Reply:
It sets up a "fake module hierarchy" on the original module for the purpose of constructing the new GraphModule. This can all go away with https://github.com/pytorch/pytorch/pull/44679 btw